src/Nethermind/Nethermind.AuRa.Test/Validators/ContractValidatorTests.cs

```diff
@@ -81,158 +90,96 @@ public void SetUp()
-                {
-                    StartBlockNumber = 100,
-                    BlockNumber = 100,
-                    InitialValidatorsCount = 1,
-                    NumberOfSteps = 1,
-                    ExpectedFinalizationCount = 1
-                };
-                
-                yield return new InitializeValidatorsTestParameters()
-                {
-                    StartBlockNumber = 1,
-                    BlockNumber = 1,
-                    InitialValidatorsCount = 2,
-                    NumberOfSteps = 2,
-                    ExpectedFinalizationCount = 0
-                };
-                
-                yield return new InitializeValidatorsTestParameters()
-                {
-                    StartBlockNumber = 1,
-                    BlockNumber = 1,
-                    InitialValidatorsCount = 2,
-                    NumberOfSteps = 3,
-                    ExpectedFinalizationCount = 1
-                };
-                
-                yield return new InitializeValidatorsTestParameters()
-                {
-                    StartBlockNumber = 1,
-                    BlockNumber = 1,
-                    InitialValidatorsCount = 3,
-                    NumberOfSteps = 2,
-                    ExpectedFinalizationCount = 0
-                };
-                
-                yield return new InitializeValidatorsTestParameters()
-                {
-                    StartBlockNumber = 1,
-                    BlockNumber = 1,
-                    InitialValidatorsCount = 3,
-                    NumberOfSteps = 3,
-                    ExpectedFinalizationCount = 1
-                };
-                
-                yield return new InitializeValidatorsTestParameters()
-                {
-                    StartBlockNumber = 1,
-                    BlockNumber = 1,
-                    InitialValidatorsCount = 10,
-                    NumberOfSteps = 5,
-                    ExpectedFinalizationCount = 0
-                };
-                
-                yield return new InitializeValidatorsTestParameters()
-                {
-                    StartBlockNumber = 1,
-                    BlockNumber = 1,
-                    InitialValidatorsCount = 10,
-                    NumberOfSteps = 7,
-                    ExpectedFinalizationCount = 1
-                };
-            }
+            var blockNumber = 10;
+            var validators = TestItem.Addresses.Take(10).ToArray();
+            var blockHash = Keccak.Compute("Test");
+            var pendingValidators = new ContractValidator.PendingValidators(blockNumber, blockHash, validators);
+            var rlp = Rlp.Encode(pendingValidators);
+            _db[ContractValidator.PendingValidatorsKey.Bytes].Returns(rlp.Bytes);
+            
+            IAuRaValidatorProcessor validator = new ContractValidator(_validator, _db, _stateProvider, _abiEncoder, _transactionProcessor, _logManager, 1);
```

Comment:
we need to store pending validators in the db? would love to hear more why it is not on the contract itself

Reply:
Two scenarios:
1. InitValidator at block N -> stop processing -> resume processing at block N+1 -> we need to load it back to state.
2. InitValidator at block N -> Reorg to N-1, we need to ignore it in that chain -> cancel Reorg -> resume processing at block N+1 at initial chain -> we need to load it back to state -> will happen with this being state db.

Reply:
just wanted to confirm that this is not stored in the contract storage?

Reply:
It generally is but contract is  not required to expose this in any API, so client has to store it itself.